processing/src/main/java/org/apache/druid/data/input/impl/AggregateProjectionSpec.java

```diff
@@ -0,0 +1,381 @@
+  public AggregateProjectionSpec(
+      @JsonProperty("name") String name,
+      @JsonProperty("virtualColumns") @Nullable VirtualColumns virtualColumns,
+      @JsonProperty("groupingColumns") @Nullable List<DimensionSchema> groupingColumns,
+      @JsonProperty("aggregators") @Nullable AggregatorFactory[] aggregators
+  )
+  {
+    this.name = name;
+    this.groupingColumns = groupingColumns == null ? Collections.emptyList() : groupingColumns;
+    this.virtualColumns = virtualColumns == null ? VirtualColumns.EMPTY : virtualColumns;
+    this.ordering = Lists.newArrayListWithCapacity(this.groupingColumns.size());
+
+    // todo (clint): maybe this initialization should be behind a supplier or something...
+    final Set<String> columnNames = new HashSet<>();
+    int foundTimePosition = -1;
+    String timeColumnName = null;
+    Granularity granularity = null;
+    // try to find the __time column equivalent, which might be a time_floor expression to model granularity
+    // bucketing. The time column is decided as the finest granularity on __time detected. If the projection does
+    // not have a time-like column, the granularity will be handled as ALL for the projection and all projection
+    // rows will use a synthetic timestamp of the minimum timestamp of the incremental index
+    for (int i = 0; i < this.getGroupingColumns().size(); i++) {
+      final DimensionSchema dimension = this.getGroupingColumns().get(i);
+      columnNames.add(dimension.getName());
+      ordering.add(OrderBy.ascending(dimension.getName()));
+      if (ColumnHolder.TIME_COLUMN_NAME.equals(dimension.getName())) {
+        timeColumnName = dimension.getName();
+        foundTimePosition = i;
+        granularity = Granularities.NONE;
+      } else {
+        final VirtualColumn vc = this.getVirtualColumns().getVirtualColumn(dimension.getName());
+        final Granularity maybeGranularity = Granularities.fromVirtualColumn(vc);
+        if (granularity == null && maybeGranularity != null) {
+          granularity = maybeGranularity;
+          timeColumnName = dimension.getName();
+          foundTimePosition = i;
+        } else if (granularity != null && maybeGranularity != null && maybeGranularity.isFinerThan(granularity)) {
+          granularity = maybeGranularity;
+          timeColumnName = dimension.getName();
+          foundTimePosition = i;
+        }
+      }
+    }
+    this.timeColumnName = timeColumnName;
+    this.timeColumnPosition = foundTimePosition;
+    this.granularity = granularity == null ? Granularities.ALL : granularity;
+
+    boolean hasCount = false;
+
+    if (aggregators != null) {
+      for (AggregatorFactory aggregator : aggregators) {
+        if (aggregator instanceof CountAggregatorFactory) {
+          hasCount = true;
+        }
+        columnNames.add(aggregator.getName());
+      }
+    }
+
+    // every projection gets a count aggregator so we can still use projections whenever a count(*) is added to a query
```

Comment:
this came out of offline discussions with @cheddar where it seemed to have big utility at relatively low cost, however thinking a bit more about it, it does seem a bit magical so i can drop it. I'll look into adding row count to the metadata so we don't need to read a number column, which side-steps the other problem this was helping solve - projections without time columns needing to know row count

Reply:
It is useful to have the `count` in there, but IMO we should address that by encouraging people to include `count` in the documentation, rather than by altering the spec provided by the user.

as an aside: I'm partly saying this because of recent efforts we've been undertaking to un-do certain transformations we've historically applied to user provided specs, like the introduction of `arrayIngestMode: array` and `forceSegmentSortByTime: false`. In those cases it's turned out that doing extra stuff on top of the spec the user provided was ultimately something we wanted to un-do.

Reply:
removed count and shuffled some stuff around as well, some of the stuff from here has been split into `AggregatorProjectionMetadata` which is what is stored in `Metadata` inside the segment instead of this API spec. This one now just finds the time column and computes ordering, and gets converted into the other thing which knows the time column and ordering already and can do less work. The metadata thing also stores the row count